src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -70,12 +72,55 @@ class RsErrorAnnotator : Annotator, HighlightRangeExtension {
             override fun visitArrayType(o: RsArrayType) = collectDiagnostics(holder, o)
             override fun visitVariantDiscriminant(o: RsVariantDiscriminant) = collectDiagnostics(holder, o)
             override fun visitPolybound(o: RsPolybound) = checkPolybound(holder, o)
+            override fun visitTryExpr(o: RsTryExpr) = checkTryExpr(holder, o)
             override fun visitTraitRef(o: RsTraitRef) = checkTraitRef(holder, o)
         }
 
         element.accept(visitor)
     }
 
+    private fun checkTryExpr(holder: AnnotationHolder, o: RsTryExpr) {
+        val items = StdKnownItems.relativeTo(o)
+        val lookup = ImplLookup.relativeTo(o)
+        val tryTrait = items.findTryTrait() ?: return
+        val fromTrait = items.findFromTrait() ?: return
+
+        val tryExprTy = o.expr.type
+        val errorTy = findErrorTyUsingTryTrait(tryExprTy, tryTrait, lookup)
+        if (errorTy == null) {
+            holder.createErrorAnnotation(
+                o.q,
+                "the `?` operator can only be applied to values that implement `std::ops::Try`"
+            )
+            return
+        }
+
+        val returnTy = findParentFnOrLambdaRetTy(o)
+        val fooErrorTy = returnTy?.let { findErrorTyUsingTryTrait(it, tryTrait, lookup) }
+        if (returnTy == null || fooErrorTy == null) {
+            holder.createErrorAnnotation(
```

Comment:
Please create error class in `RsDiagnostics` and use it (see `checkTraitRef` for example)